Skia support, multi-editors, monaco updates

clairernovotny opened the pull request:
This pull request introduces several significant improvements and updates across the codebase, focusing on modernizing project configuration, improving nullability and event handling, updating dependencies and build tooling, and enhancing the code editor's initialization logic. The changes aim to make the project more robust, maintainable, and compatible with newer tooling.

**Project structure and configuration modernization:**

* Added `Directory.Build.props`, `Directory.Build.targets`, and `Directory.Packages.props` files to enable central management of build properties and package versions, including enabling implicit usings, nullable reference types, treating warnings as errors, and suppressing specific warnings (`NU1507`, `NETSDK1201`, `PRI257`, `CS1998`). [[1]](diffhunk://#diff-9da24614831c308827a1ae533ffea392c97638c261dd42bd0f5226baa136d16eR1-R16) [[2]](diffhunk://#diff-50e91c82311ea26f2a73202525dfdf2b0a89c178ee666b2bf3ad4c84ac4c06dcR1-R2) [[3]](diffhunk://#diff-5baf5f9e448ad54ab25a091adee0da05d4d228481c9200518fcb1b53a65d4156R1-R17)
* Updated the solution file (`MonacoEditorComponent.sln`) and introduced a solution filter (`MonacoEditorComponent.slnf`) to improve project organization and streamline loading in Visual Studio. [[1]](diffhunk://#diff-cb55398e0df2a6bd220277f63ede12f0bb7d1de9d92c480333fb09b83e3279c9R12-R22) [[2]](diffhunk://#diff-e111ea40a2ae34fa231191926ffb92baebfd78f9fd0b313a2c9183073e1a12acR1-R10) [[3]](diffhunk://#diff-cb55398e0df2a6bd220277f63ede12f0bb7d1de9d92c480333fb09b83e3279c9L83)

**Dependency and tooling updates:**

* Upgraded `monaco-editor`, `typedoc`, and `typescript` dependencies to more recent versions in `GenerateMonacoTypings/package.json`. Also updated the TypedocConverter tool to v2.6 and adjusted the PowerShell script to use new options for type conversion. [[1]](diffhunk://#diff-59ecc8bfd76edfe90fcd8522ed71dc3bb9ab3e7ef653f5f8714aea8fd99303e4L13-R20) [[2]](diffhunk://#diff-0b17b8b1f5117ea3dba5c20aa5320feee06fae0c9b64dd88e6d7835f51ae3c4cR38) [[3]](diffhunk://#diff-0b17b8b1f5117ea3dba5c20aa5320feee06fae0c9b64dd88e6d7835f51ae3c4cL50-R51)

**Build process enhancements:**

* Modified the CI pipeline configuration (`.vsts-ci.yml`) to use the new solution filter file, added binary logging to MSBuild, and ensured artifacts are always published regardless of build success. [[1]](diffhunk://#diff-fb7d8d37fd9e28ca688fd102422af3e1d1df13a50d7f2deb2e145aa5ef449057L21-R22) [[2]](diffhunk://#diff-fb7d8d37fd9e28ca688fd102422af3e1d1df13a50d7f2deb2e145aa5ef449057R37)

**Code editor improvements and nullability:**

* Refactored event declarations in `CodeEditor.Events.cs` to use nullable event handlers and renamed events for clarity (e.g., `EditorLoading`, `EditorLoaded`). Improved null safety throughout the file, updated initialization logic, and enhanced focus management and theme handling to be more robust and compatible with nullable reference types. [[1]](diffhunk://#diff-dbe8115f75a9a5cc01bf6919573e952219363c4a96c2d52f32c3fb18f7cae045R8) [[2]](diffhunk://#diff-dbe8115f75a9a5cc01bf6919573e952219363c4a96c2d52f32c3fb18f7cae045L22-L114) [[3]](diffhunk://#diff-dbe8115f75a9a5cc01bf6919573e952219363c4a96c2d52f32c3fb18f7cae045R133-R193) [[4]](diffhunk://#diff-dbe8115f75a9a5cc01bf6919573e952219363c4a96c2d52f32c3fb18f7cae045L162-R205) [[5]](diffhunk://#diff-dbe8115f75a9a5cc01bf6919573e952219363c4a96c2d52f32c3fb18f7cae045R229-L196)
* Updated the `CommandHandler` delegate in `CodeEditor.Methods.cs` to accept nullable parameters, aligning with the project's move to nullable reference types.

These changes collectively modernize the project, improve safety and maintainability, and ensure compatibility with current .NET and JavaScript tooling.

jeromelaban committed:
wip

jeromelaban committed:
wip

jeromelaban committed:
wip

jeromelaban committed:
wip

jeromelaban committed:
nullable

jeromelaban committed:
more nullable

jeromelaban committed:
more fixes

jeromelaban committed:
chore: Adjust format

jeromelaban committed:
Update to use latest uno 6 templates

jeromelaban committed:
update to uno latest

jeromelaban committed:
chore: Add missing file

clairernovotny committed:
Update dependency versions

clairernovotny committed:
Update to latest Monaco (0.52.0)

clairernovotny committed:
Auto add and activate tab

clairernovotny committed:
NullReferenceException repro

MartinZikmund committed:
Add missing styles

clairernovotny committed:
Add resources for 6.3

clairernovotny committed:
Don't refresh layout until editor is fully loaded

clairernovotny committed:
use private sdk

clairernovotny committed:
add simple test page

clairernovotny committed:
Update to latest sdks

clairernovotny committed:
Update Uno.WinUI version and add net10.0 support

Updated Uno.WinUI package version in Directory.Packages.props to 6.4.0-dev.212.
Enhanced MonacoEditorComponent.csproj to support net10.0 framework alongside net9.0,
including updates to target frameworks, conditional property groups, and item groups.
Modified App.xaml.cs to navigate to MainPage by default.
Added BlazorWebAssemblyJiterpreter property in MonacoEditorTestApp.csproj and
generalized file inclusion pattern for browserwasm.

clairernovotny committed:
Update to Monaco Editor 0.54

clairernovotny committed:
Refactor and cleanup project files

clairernovotny committed:
Update Uno dependencies and project configurations

Updated the `Uno.WinUI` package to `6.4.0-dev.235` and the `Uno.Sdk` versions in `global.json` to `6.4.0-dev.50` and `6.4.0-dev.256`.

Changed the project SDK in `MonacoEditorTestApp.csproj` from `Uno.Sdk.Private` to `Uno.Sdk` and upgraded the target framework to `net10.0-browserwasm`.

Commented out unused embedded resources in `MonacoEditorComponent.csproj` and added the `<WasmFingerprintAssets>` property with a value of `false` to optimize asset handling.

clairernovotny committed:
Properly include all of the monaco files

clairernovotny committed:
Switch to Nerdbank.GitVersioning and update metadata

Modernized versioning by replacing `gitversion.yml` with `version.json` using the `Nerdbank.GitVersioning` schema. Added a global package reference for `Nerdbank.GitVersioning` (v3.8.118) in `Directory.Packages.props`.

Updated `MonacoEditorComponent.csproj`:
- Changed `Title` to "Monaco Editor Component WASM".
- Enhanced NuGet metadata with `Authors`, `Copyright`, `PackageLicenseExpression`, and `PackageTags`.
- Simplified packaging by removing `NuspecProperties` and `NuspecFile`.
- Excluded unnecessary files from the NuGet package using `Pack="false"`.

Improved project structure and clarified the scope of the Monaco Editor component.

clairernovotny committed:
Improve build process and packaging rules

- Added `<ContinuousIntegrationBuild>` property for CI builds.
- Updated copyright year to dynamically use the current year.
- Adjusted content packaging rules to include necessary files.
- Introduced framework-specific content packaging logic.
- Improved handling of primary target framework extraction.
- Allowed `.pdb` files in package build output.
- Removed deprecated `<GenerateLibraryLayout>` property.
- General cleanup of content inclusion and resource declarations.

clairernovotny committed:
Update CI pipeline and simplify solution structure

- Updated CI pipeline to use `windows-latest` VM image.
- Migrated to `.NET 10 SDK` and added workload installation.
- Replaced `MSBuild` task with `DotNetCoreCLI` for builds.
- Introduced `gitversion-run.yml` for GitVersion setup.
- Removed legacy `gitversion.yml` and Windows SDK install step.
- Simplified solution structure with new `.slnx` format.
- Improved build artifact management for version variables.
- Cleaned up unused configurations and legacy files.

jeromelaban commented on MonacoEditorComponent/CodeEditor/CodeEditor.cs:
```diff
@@ -88,33 +127,73 @@ private async void Options_PropertyChanged(object sender, PropertyChangedEventAr
+                ReadOnly = Options.ReadOnly.Value;
+            }
+
+            Console.WriteLine($"CodeEditor_Loaded [{_model}] [{_view}] ({GetHashCode():x8})");
```
We should most this to logs

jeromelaban reviewed (commented):
Some comments

clairernovotny committed:
Code cleanup

clairernovotny committed:
Refactor: Modernize codebase with C# 9+ features

- Updated `.editorconfig` to enforce consistent formatting and suppress specific diagnostics.
- Enabled `LangVersion` to `latest` in `Directory.Build.props`.
- Replaced `Array.Empty<T>()` with `[]` and used target-typed `new()` for object initializations.
- Simplified null checks using `is not` syntax and null-conditional operators.
- Refactored `switch` statements to `switch` expressions for better readability.
- Marked obsolete constructors and improved method argument handling with `[..]`.
- Reorganized `using` directives and improved code readability across multiple files.
- Added `.editorconfig` to the project content in `MonacoEditorComponent.csproj`.

clairernovotny committed:
Added a call to `NativeMethods.RefreshLayout` after `NativeMethods.InitializeMonaco` to ensure proper layout refresh during Monaco editor initialization.

clairernovotny committed:
Enhance editor API with documentation and new features

Added XML documentation to improve code readability and usability.
Introduced new properties and methods across multiple classes to
extend functionality, including `ContextKey`, `FindMatch`,
`IRelatedInformation`, `CssGlyphStyle`, `IMarkdownString`,
`CodeAction`, `CodeLens`, `CompletionList`, `TextEdit`,
`WorkspaceEdit`, and `WorkspaceTextEdit`.

Ensured `WorkspaceTextEdit.ModelVersionId` is nullable to
prevent serialization issues. Added extension methods for
`ICssStyle` and `IMarkdownString` to simplify CSS wrapping
and Markdown conversion. Improved maintainability with
detailed summaries for classes, properties, and methods.

clairernovotny committed:
Make CodeEditor properties public for external access

clairernovotny committed:
restore generatelibrarylayout

clairernovotny committed:
Refactor MonacoEditorComponent project file

Refined content inclusion and packaging configurations:
- Updated `<IncludeContentInPack>` logic to use `_PrimaryTfm`.
- Adjusted `<ItemGroup>` to better handle TypeScript and content files.
- Replaced `<Content>` with `<None>` for `CodeEditor.html`.
- Added specific inclusions for Monaco Editor files.
- Removed unused embedded resource `MonacoEditorComponent.rd.xml`.
- Commented out TFM-specific content `<PropertyGroup>`.
- Added `<Target>` for TFM-specific Monaco content packaging.

clairernovotny committed:
Update packaging behavior and TFM handling

Commented out `<IncludeContentInPack>` to potentially allow content files in the package.

clairernovotny committed:
Set IncludeContentInPack to false in project file

Uncommented and set the `<IncludeContentInPack>` property to
`false` to ensure content files are excluded from the NuGet
package during the build process. This change prevents
unnecessary content from being included in the package.

clairernovotny committed:
Added `<clear />` tags to `<packageSources>` to remove any pre-existin sources.

Switch CI to Ubuntu, update build props, and nuget config

Updated the CI pipeline to use `ubuntu-latest` instead of
`windows-latest` for the `Windows` job. Added the
`<UseArtifactsOutput>` property in `Directory.Build.props` to
enable artifact output.

clairernovotny committed:
Set name to Build

clairernovotny committed:
Simplify artifact publishing and fix variable casing

clairernovotny committed:
fix case

carldebilly commented on changelog.md:
Is this file still relevant?

carldebilly commented on global.json:
```diff
@@ -1,5 +1,9 @@
   "msbuild-sdks": {
-    "MSBuild.Sdk.Extras": "2.1.2"
+    "Uno.Sdk": "6.4.0-dev.50",
+    "Uno.Sdk.Private": "6.4.0-dev.256"
```
Usually we don't use both in a same solution or it would cause strange problems with the DevServer.

I suggestion keeping only `Uno.Sdk.Private` here.

carldebilly replied:
Note: [`Uno.Sdk 6.4.0-dev.50` uses Uno version `6.4.0-dev.235`](https://www.nuget.org/packages/Uno.Sdk/6.4.0-dev.50#readme-body-tab), not `6.4.0-dev.256`

clairernovotny committed:
Refine artifact publishing and add dynamic source root

Updated `.vsts-ci.yml` to adjust artifact publishing paths:
- Changed publish path to `package/release` for better structure.
- Removed redundant log publishing path.

Modified `Directory.Build.targets` to include a `SourceRoot` item:
- Dynamically includes `$(ArtifactsPath)` if the variable is set.
- Supports improved build and packaging workflows.

clairernovotny committed:
Ensure directory path in SourceRoot Include attribute

Updated the `<SourceRoot>` element in `Directory.Build.targets`
to append a trailing slash (`/`) to the `$(ArtifactsPath)`

carldebilly reviewed (commented):
LGTM

clairernovotny committed:
added copilot setup

clairernovotny committed:
set sdk ver

Copilot committed:
Apply initial property values on CodeEditor load to fix initialization timing and syntax highlighting (#29)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/unoplatform/uno.monaco-editor-uwp?pullRequest=28) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/unoplatform/uno.monaco-editor-uwp?pullRequest=28) before we can accept your contribution.<br/>**3** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: jeromelaban<br/>:white_check_mark: clairernovotny<br/>:white_check_mark: MartinZikmund<br/>:x: Copilot<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/unoplatform/uno.monaco-editor-uwp?pullRequest=28) it.</sub>

clairernovotny committed:
Add FileExtension property to infer code language

Introduced a `FileExtension` property in the `CodeEditor` class to automatically determine the code language based on file extensions. This is backed by a new `DependencyProperty` and integrates with a JavaScript function `languageIdFromFileName` for language mapping using Monaco Editor.

- Added `LanguageIdFromFileName` method in `NativeMethods` for JS interop.
- Implemented `languageIdFromFileName` in `asyncCallbackHelpers.ts`.
- Updated `globalThis` to include the new function.
- Refactored `using` directives for consistency and removed redundancies.
- Added XML documentation for the `FileExtension` property.
- Improved maintainability and readability across related files.

clairernovotny committed:
Make DebugLogger accessible to JavaScript in UWP apps

Added the `AllowForWeb` attribute to enable JavaScript access
to the `DebugLogger` class in UWP apps. Suppressed the CA1822
warning for the `Log` method to avoid marking it as static,
ensuring compatibility with the existing design. Reformatted
the `#if DEBUG` block for improved readability and consistency.

clairernovotny committed:
Refactor language detection by file extension

Replaced `FileExtension` property in `CodeEditor` with a new approach that determines code language based on file extensions. Removed the `LanguageIdFromFileName` method and introduced `LanguageIdFromExtension` for improved efficiency and clarity.

- Removed `FileExtension` property and its `DependencyProperty` logic.
- Updated `globalThis.languageIdFromFileName` to `languageIdFromExtension` in JavaScript.
- Modified `LanguagesHelper` to include `GetCodeLanguageFromExtension` method.
- Changed `LanguagesHelper` class to `sealed partial` for extensibility.
- Adjusted imports and removed unused namespaces for cleanup.
- Updated comments and attributes to reflect the new design.

These changes simplify the codebase, reduce redundancy, and align with best practices for language detection.

clairernovotny committed:
Update JS import for LanguageIdFromExtension method

The `JSImport` attribute for the `LanguageIdFromExtension`
method in the `NativeMethods` class was updated to reference
the `languageIdFromExtension` function from `globalThis`
instead of `languageIdFromFileName`. This change reflects a
shift in functionality to determine the language ID based on
file extensions rather than file names.

clairernovotny committed:
Comment out debug logs in asyncCallbackHelpers.ts

This commit comments out multiple `console.debug` statements
throughout the `asyncCallbackHelpers.ts` file. These logs were
used for tracking various stages of the Monaco Editor
initialization process, including grabbing options, creating
the editor, listening for changes, and setting themes.

The changes aim to reduce console output for cleaner logs in
production or to avoid unnecessary clutter during runtime.
No functional code changes were made.

clairernovotny committed:
Update Uno and SDK versions to stable releases

Updated `Directory.Packages.props` to use `Uno.WinUI` version `6.4.185`, transitioning from a development version (`6.4.0-dev.235`) to a stable release.

In `global.json`:
- Updated `Uno.Sdk` to `6.4.31` and `Uno.Sdk.Private` to `6.4.185`, moving from development versions to stable releases.
- Removed `allowPrerelease` to disallow prerelease SDKs.
- Updated .NET SDK version from `10.0.100-rc.2.25502.107` (release candidate) to `10.0.100` (stable).

clairernovotny committed:
Update CI pipeline and remove NuGet packaging config

Updated `.vsts-ci.yml` to use .NET 10 SDK and added a script to install WebAssembly tools for .NET 9. Removed the `gitversion-run.yml` template, indicating a shift in versioning strategy.

Deleted `MonacoEditorComponent.nuspec` and its reference in the solution file, discontinuing or restructuring the NuGet packaging process for the `Uno.Monaco.Editor` component. These changes modernize the build pipeline and adjust the distribution strategy.
